src/xenia/kernel/friend_presence_object_json.cc

```diff
@@ -0,0 +1,201 @@
+  }
+
+  if (obj.HasMember("richPresence")) {
+    const std::u16string richPresence =
+        xe::to_utf16(obj["richPresence"].GetString());
+
+    RichPresence(richPresence);
+  }
+
+  return true;
+}
+
+bool FriendPresenceObjectJSON::Serialize(
+    rapidjson::PrettyWriter<rapidjson::StringBuffer>* writer) const {
+  // writer->StartObject();
+
+  // writer->String("xuid");
+  // writer->String(fmt::format("{:016X}", xuid_));
+
+  // writer->String("gamertag");
+  // writer->Uint(gamertag_);
+
+  // writer->String("state");
+  // writer->Uint(state_);
+
+  // writer->String("sessionId");
+  // writer->String(fmt::format("{:016X}", sessionId_));
+
+  // writer->String("titleId");
+  // writer->String(title_id_);
+
+  // writer->String("stateChangeTime");
+  // writer->Uint64(state_change_time_);
+
+  // writer->String("richPresenceStateSize");
+  // writer->Uint(rich_state_presence_size_);
+
+  //// Problem :/
+  // writer->String("richPresence");
+  //// writer->String(rich_presence_);
+
+  // writer->EndObject();
+
+  return false;
+}
+
+FriendsPresenceObjectJSON::FriendsPresenceObjectJSON() : xuids_(0) {}
+
+FriendsPresenceObjectJSON::~FriendsPresenceObjectJSON() {}
+
+bool FriendsPresenceObjectJSON::Deserialize(const rapidjson::Value& obj) {
+  if (!obj.IsArray()) {
+    return false;
+  }
+
+  for (const auto& presenceObj : obj.GetArray()) {
+    FriendPresenceObjectJSON* presence = new FriendPresenceObjectJSON();
+    presence->Deserialize(presenceObj.GetObj());
+
+    players_presence_.push_back(*presence);
+  }
+
+  return true;
+}
+
+bool FriendsPresenceObjectJSON::Serialize(
+    rapidjson::PrettyWriter<rapidjson::StringBuffer>* writer) const {
+  writer->StartObject();
+  writer->String("xuids");
+
+  writer->StartArray();
+
+  for each (const auto xuid in xuids_) {
+    writer->String(fmt::format("{:016X}", xuid));
+  }
+
+  writer->EndArray();
+
+  writer->EndObject();
+
+  return true;
+}
+
+X_ONLINE_PRESENCE FriendPresenceObjectJSON::ToOnlineRichPresence() const {
+  X_ONLINE_PRESENCE presence = {};
+
+  presence.xuid = XUID();
+  presence.state = State();
+  std::memcpy(&presence.session_id, &SessionID(), sizeof(XNKID));
+
+  if (!TitleID().empty()) {
+    presence.title_id = string_util::from_string<uint32_t>(TitleID(), true);
+  }
+
+  presence.state_change_time = StateChangeTime();
+  presence.cchRichPresence = RichStatePresenceSize();
+
+  const std::u16string presence_string = RichPresence();
+
+  char16_t* rich_presence_ptr =
+      reinterpret_cast<char16_t*>(presence.wszRichPresence);
+  xe::string_util::copy_and_swap_truncating(rich_presence_ptr, presence_string,
+                                            presence.cchRichPresence);
+
+  return presence;
+}
+
+X_ONLINE_FRIEND FriendPresenceObjectJSON::GetFriendPresence() const {
+  X_ONLINE_FRIEND peer = {};
+
+  peer.xuid = XUID();
+
+  char* gamertag_ptr = reinterpret_cast<char*>(peer.Gamertag);
+  strcpy(gamertag_ptr, Gamertag().c_str());
```

Comment:
I would limit it to 15 chars as gamertag should be limited. If gamertag won't be limited then it can overwrite data after that pointer